packages/gpl-core/src/badge.ts

```diff
@@ -0,0 +1,314 @@
+  address: string;
+  issuer: string;
+  holder: string;
+  update_authority: string;
+  schema: string;
+  metadata_uri: string;
+  refreshed_at?: Date;
+  slot_created_at?: Date;
+  slot_updated_at?: Date;
+  created_at?: Date;
+}
+
+export interface GraphQLIssuer {
+  address: string;
+  authority: string;
+  verified: boolean;
+  refreshed_at?: Date;
+  slot_created_at?: Date;
+  slot_updated_at?: Date;
+  created_at?: Date;
+}
+
+export class Badge {
+  private readonly sdk: SDK;
+
+  constructor(sdk: SDK) {
+    this.sdk = sdk;
+  }
+
+  public async getBadge(userAccount: anchor.web3.PublicKey) {
+    return await this.sdk.program.account.badge.fetch(userAccount);
+  }
+
+  public async getOrCreateBadge(
+    metadataUri: string,
+    issuer: anchor.web3.PublicKey,
+    schema: anchor.web3.PublicKey,
+    holder: anchor.web3.PublicKey,
+    updateAuthority: anchor.web3.PublicKey,
+    authority: anchor.web3.PublicKey,
+  ) {
+    const [badgePDA, _] = anchor.web3.PublicKey.findProgramAddressSync(
+      [
+        Buffer.from("badge"),
+        issuer.toBuffer(),
+        schema.toBuffer(),
+        holder.toBuffer(),
+      ],
+      this.sdk.program.programId
+    );
+
+    try {
+      await this.getBadge(badgePDA);
+    } catch (err) {
+      await (await this.createBadge(metadataUri, issuer, schema, holder, updateAuthority, authority)).instructionMethodBuilder.rpc();
+    }
+    return badgePDA;
+  }
+
+  public async createBadge(
+    metadataUri: string,
+    issuer: anchor.web3.PublicKey,
+    schema: anchor.web3.PublicKey,
+    holderProfile: anchor.web3.PublicKey,
+    updateAuthority: anchor.web3.PublicKey,
+    authority: anchor.web3.PublicKey) {
+    const instructionMethodBuilder = this.sdk.program.methods
+      .createBadge(metadataUri)
+      .accounts({
+        issuer,
+        schema,
+        holder: holderProfile,
+        updateAuthority,
+        authority,
+      });
+    const pubKeys = await instructionMethodBuilder.pubkeys();
+    const badgePDA = pubKeys.badge as anchor.web3.PublicKey;
+    return {
+      instructionMethodBuilder,
+      badgePDA,
+    };
+  }
+
+  public async updateBadge(
+    metadataUri: string,
+    badgeAccount: anchor.web3.PublicKey,
+    issuer: anchor.web3.PublicKey,
+    schema: anchor.web3.PublicKey,
+    signer: anchor.web3.PublicKey
+  ) {
+    return this.sdk.program.methods
+      .updateBadge(metadataUri)
+      .accounts({
+        badge: badgeAccount,
+        issuer,
+        schema,
+        signer,
+      });
+  }
+
+  public async burnBadge(
+    badgeAccount: anchor.web3.PublicKey,
+    issuer: anchor.web3.PublicKey,
+    schema: anchor.web3.PublicKey,
+    holder: anchor.web3.PublicKey,
+    signer: anchor.web3.PublicKey
+  ) {
+    return this.sdk.program.methods
+      .burnBadge()
+      .accounts({
+        badge: badgeAccount,
+        issuer,
+        schema,
+        holder,
+        signer,
+      });
+  }
+
+  public async getOrCreateIssuer(
```

Comment:
Same as badge, I don't think we should encourage this. Do you think of any scenarios where this might be necessary?